magres_old parsing improvements

jkshenton opened the pull request:
I've added in something to parse extra information in magres_old blocks. A lot of it will be redundant information (i.e. also in the regular magres block), but the hyperfine tensors, for example, aren't written out in the magres blocks, only magres_old. 

Done:

- output from magres_old blocks now mirrors standard magres block
-  magres_old can now parse
   - ms
   - efg
   - isc_*
   - hf (hyperfine tensors)
- updated test magres file to include new quantities in magres_old
- changed the structure of the tensors in magres block to ThreeByThreeMatrix to remove any ambiguity in ordering (row vs col)
- Updated test json and yaml files to the new structure



Optional things left to do:

- ~~[ ] Add in summaries of tensors~~ 
Edit: Ignoring this for now/downstream packages

For now I removed the parsing of the eigenvectors/values and the isotropy etc. for each tensor - these could be added back in, but then maybe a structure like:

```
- magres_old
    - magnetic_shielding | electric_field_gradient | hyperfine | j_coupling
        - atomindex | atomindex1,atomindex2
            - tensor
            - eigenvalues
            - eigenvectors
            - isotropy
            - anisotropy
            - asymmetry
```       

- [x] Combine magres and magres_old data to remove redundancy

jkshenton committed:
Magres parser improvements

- output from magres_old blocks now mirrors standard magres block
-  magres_old can now parse
   - ms
   - efg
   - isc_*
   - hf (hyperfine tensors)
- updated test magres file to include new quantities in magres_old
- changed the structure of the tensors in magres block to ThreeByThreeMatrix to remove any ambiguity in ordering (row vs col)
- Updated test json and yaml files to the new structure

jkshenton committed:
Magres parser improvements

- output from magres_old blocks now mirrors standard magres block
-  magres_old can now parse
   - ms
   - efg
   - isc_*
   - hf (hyperfine tensors)
- updated test magres file to include new quantities in magres_old
- changed the structure of the tensors in magres block to ThreeByThreeMatrix to remove any ambiguity in ordering (row vs col)
- Updated test json and yaml files to the new structure

jkshenton committed:
Linting fixes

jkshenton committed:
Linting fixes

oerc0122 commented on castep_outputs/utilities/castep_res.py:
```diff
@@ -497,6 +497,50 @@ def get_atom_parts(spec: str) -> dict[str, str]:
+    ),
+    # Magnetic shielding tensor
+    "ms_tensor": re.compile(
+        r"\s{0,}(.*?) Shielding Tensor[\r\n]+"
+        r"\s+([0-9\.\-]+)\s+([0-9\.\-]+)\s+([0-9\.\-]+)[\n\r]+"
+        r"\s+([0-9\.\-]+)\s+([0-9\.\-]+)\s+([0-9\.\-]+)[\n\r]+"
+        r"\s+([0-9\.\-]+)\s+([0-9\.\-]+)\s+([0-9\.\-]+)\s+",
```
```suggestion
TENSOR_RE = (r"{labelled_floats(('xx', 'xy', 'xz')}[\n\r]+"
             r"{labelled_floats(('yx', 'yy', 'yz')}[\n\r]+"
             r"{labelled_floats(('zx', 'zy', 'zz')}")
...

        rf"\s*(.*?) Shielding Tensor[\r\n]+{TENSOR_RE}"
```

oerc0122 replied:
This could then be reused for other tensors.

oerc0122 reviewed (changes requested):
Thanks for this, starting to look a lot better! Just some style questions and hopefully simplified RE structure.

oerc0122 commented on castep_outputs/parsers/magres_file_parser.py:
```diff
@@ -223,33 +255,114 @@ def _process_magres_old_block(
+    data["atoms"]["units"]["atom"] = "Angstrom"
+
+    found_atoms = set()
+    coords = REs.MAGRES_OLD_RE["coords"].findall(str(block))
```
Actually, I just realised you're matching against the whole "file", where that won't necessarily help. I will probably convert this to an iterated one at some point, just for consistency with other readers, but this is good progress and very helpful.

jkshenton replied:
I first tried to use the `Block.from_re` but didn't get it to work so I just borrowed from here:
https://github.com/tfgg/magres-format/blob/master/magres/oldmagres.py

An iterated version would be great!

jkshenton commented on castep_outputs/parsers/magres_file_parser.py:
```diff
@@ -59,8 +86,27 @@ class MagresInfo(TypedDict):
                    float, float, float,
                    float, float, float]
-    #: Magnetic susceptibility tensor.
+    # Magnetic shielding tensor.
```
I didn't notice the `:` - yes I think they do want to be document labels. I'll do that

jkshenton committed:
Simplify magres old RegEx

jkshenton committed:
Simplify magres old RegEx

jkshenton committed:
magres old parser style fixes

jkshenton committed:
magres old parser style fixes

jkshenton committed:
ruff linting fixes for magres parser

jkshenton committed:
ruff linting fixes for magres parser

jkshenton commented:
Thanks for the comments and suggestions @oerc0122 , I think I've addressed all of them now.

Do you have thoughts on
1. merging redundant data (much of the magres_old block has the same info as the magres block, so it could be used to sanity-check, for example. I mainly to parse the `magres_old` block for the hyperfine tensors (which only appear in the `magres_old`, not in the `magres` one - for now...). One option could be to add in the missing data into a single magres output (rather than having magres and magres_old), though that might violate the design of this library (?)
2. We could add in the eigenvectors/values and isotropy etc. - these are redundant in that you can compute these quantities based on the tensors that are parse, but they might help people debug conventions etc...?

oerc0122 commented:
> Thanks for the comments and suggestions @oerc0122 , I think I've addressed all of them now.
> 
> Do you have thoughts on
> 
> 1. merging redundant data (much of the magres_old block has the same info as the magres block, so it could be used to sanity-check, for example. I mainly to parse the `magres_old` block for the hyperfine tensors (which only appear in the `magres_old`, not in the `magres` one - for now...). One option could be to add in the missing data into a single magres output (rather than having magres and magres_old), though that might violate the design of this library (?)
> 2. We could add in the eigenvectors/values and isotropy etc. - these are redundant in that you can compute these quantities based on the tensors that are parse, but they might help people debug conventions etc...?

I think that duplicating data is probably unnecessary. 

Ultimately, I'm not sure we want the `magres`, `magres_old` separation. It was there because that was the file structure, but I think we just want the `magres` dictionary as a whole as that's all the useful information in the file, to that end, I think adding the eigenvectors/value, etc. would be a good thing. If people don't want them they can not use them quite easily, but if information's not there, they can't.

oerc0122 reviewed (dismissed):
Depending on your thoughts and what you're using, we can leave the merging of `magres` and `magres_old` until the refactor to an iterative form. It will, however, be an API breaking change.

oerc0122 commented:
@jkshenton 
I was wondering whether you had any more thoughts on the PR (since you have some unresolved ones), or whether you're happy as is. 

We can get it merged now and sort out some of the details later or you can work on it, or I can take over and do some of the cleanup you talked about. 

I think this would be a good thing to get in.

jkshenton commented:
@oerc0122 apologies for this dropping off the radar a bit!

If you'd be happy to take on the merging of the magres_old and magres data, I would be very grateful! If not, I can have a go, but probably only in 2 weeks' time.

Happy to leave the parsing/computing of eigenvectors/values and other derived quantities to other programs/a future PR.

oerc0122 committed:
Merge magres_old with magres

oerc0122 committed:
Merge magres_old with magres

oerc0122 commented:
@jkshenton Do you mind checking if this meets your needs or if I've inadvertently removed useful information? I've never done much with NMR so just want to make sure!

Commit:
Add deep merge function for magres and update test data files

- Implemented a recursive dictionary merge function `deep_merge_dict` to facilitate merging nested dictionaries.
- Enhanced `test.magres` files with additional data entries.
- ions:lattice now returns a 3x3 instead of a 1x9
- Removed aliasing - keeping ms, efg, isc and hf as users will expect. I'm not sure about this, but I think this would be the least surprising option for users
- Removed efg_local and efg_nonlocal following discussions with domain experts
- Added hf to main magres block (should be added in future versions of CASTEP magres output)

Commit:
Add deep merge function for magres and update test data files

- Implemented a recursive dictionary merge function `deep_merge_dict` to facilitate merging nested dictionaries.
- Enhanced `test.magres` files with additional data entries.
- ions:lattice now returns a 3x3 instead of a 1x9
- Removed aliasing - keeping ms, efg, isc and hf as users will expect. I'm not sure about this, but I think this would be the least surprising option for users
- Removed efg_local and efg_nonlocal following discussions with domain experts
- Added hf to main magres block (should be added in future versions of CASTEP magres output)

jkshenton commented:
@oerc0122 I've just got round to it and made some additional changes:

I think the `|` merging missed the nested dictionary items, so I've added a nested merge - is that correct or is there a better way? 

I also added stuff to the test.magres so it covers more cases.

The json, ruamel, and pyyaml files I generated have a bit different format in places to the previous versions - not sure why that is

oerc0122 reviewed (dismissed):
Looks pretty sensible to me. Of course since I've technically contributed might want @ajjackson 's review before merging.

oerc0122 commented:
Turned out to just be a merge error, thankfully. Regenerating the file removed the duplication.

Commit:
Make the deep_merge_dict use deepcopy for the values. Also regenerate the test magres.pyyaml

Commit:
Make the deep_merge_dict use deepcopy for the values. Also regenerate the test magres.pyyaml

oerc0122 reviewed (dismissed):
I'm happy with this if @ajjackson is.

jkshenton commented on castep_outputs/parsers/magres_file_parser.py:
```diff
@@ -170,58 +228,40 @@ def _process_magres_block(block: Block, version: int) -> dict[str, str | ThreeBy
             accum["units"][key] = val
 
-        elif (words := line.split())[0] in {"ms", "efg", "efg_local", "efg_nonlocal"}:
+        elif (words := line.split())[0] in {"ms", "efg", "hf"}:
```
After some earlier discussions with @jryates, I would say that it's completely safe to drop parsing support for `efg_local` and `efg_nonlocal` as they're not useful outside some extremely niche use cases and would instead only add to confusion (and they haven't been written out by CASTEP for a while now).

ajjackson commented on castep_outputs/parsers/magres_file_parser.py:
```diff
@@ -232,34 +272,144 @@ def _process_magres_old_block(
+
+    # For any isc tensor tags, add the units if not present
+    for tag in ("isc", "isc_fc", "isc_spin", "isc_orbital_p", "isc_orbital_d"):
+        if tag in data["magres"] and tag not in data["magres"]["units"]:
```
Do we actually expect these to be present sometimes and not others?

Does anything bad happen if we remove the `and` branch and just have `if tag in data["magres"]` here?

oerc0122 committed:
Respond to review

oerc0122 committed:
Respond to review

oerc0122 committed:
Rework magres slightly
